python/setup_project/storage_map_page.py

```diff
@@ -0,0 +1,398 @@
+        for (root_name, root_info) in self._required_roots:
+
+            # create the widget and set all the values
+            map_widget = StorageMapWidget(self._storages_model)
+            map_widget.root_name = root_name
+            map_widget.root_info = root_info
+            map_widget.set_count(count, num_roots)
+
+            # set the best guess for this mapping based on saved historical
+            # mappings (this could be None)
+            map_widget.best_guess = self._historical_mappings.get(root_name)
+
+            # add the map widget to the ui
+            layout.addWidget(map_widget)
+
+            # get alerts when a storage is saved for this mapping
+            map_widget.storage_saved.connect(self._on_storage_saved)
+
+            # keep a reference to the map widget
+            self._map_widgets.append(map_widget)
+
+            # tell the widget to make an educated guess as to which storage
+            # should be used for the supplied root name/info.
+            map_widget.guess_storage()
+
+            count += 1
+
+        layout.addStretch()
+
+        # clear out any errors
+        ui.storage_errors.setText("")
+
+    def add_mapping(self, root_name, root_info):
+        """Add a new storage mapping widget to the list.
+
+        This should be called before the page is initialized.
+        """
+        self._required_roots.append((root_name, root_info))
+
+    def clear_roots(self):
+        """Clear any required roots already set for the page.
+
+        This is typically called when the wizard's "back" button is used. The
+        roots are cleared so they can be rebuilt with potentially a different
+        configuration.
+        """
+
+        # the wizard instance and its UI
+        wiz = self.wizard()
+        ui = wiz.ui
+
+        # the scroll area widget where the mapping widgets live
+        layout = ui.storage_map_area_widget.layout()
+
+        # reset the internal data
+        self._required_roots = []
+        self._map_widgets = []
+
+        # iterate over all layout items and remove them. if the item is a
+        # storage map widget, delete it.
+        for index in range(0, layout.count()):
+            layout_item = layout.takeAt(0)
+            widget = None
+            if layout_item:
+                widget = layout_item.widget()
+            layout.removeItem(layout_item)
+            if widget and isinstance(widget, StorageMapWidget):
+                widget.deleteLater()
+
+    def set_config_uri(self, uri):
+        """Set the config URI."""
+        self._uri = uri
+
+    def validatePage(self):
+        """The 'next' button was pushed. See if the mappings are valid."""
+
+        # the wizard instance and its UI
+        wiz = self.wizard()
+        ui = wiz.ui
+
+        # clear any errors
+        ui.storage_errors.setText("")
+
+        # get the path key for the current os
+        if sys.platform.startswith("linux"):
```

Comment:
Can you use a [`ShotgunPath`](http://developer.shotgunsoftware.com/tk-core/utils.html#shotgunpath) here? Maybe also in other places?